ddp-workspace/projects/ddp-dsm-ui/src/app/tags/cohort-tag/cohort-tag.component.ts

```diff
@@ -0,0 +1,77 @@
+import { Component, Input, OnChanges, OnInit, SimpleChanges } from '@angular/core';
+import { CohortTag } from './cohort-tag.model';
+import {MatChipInputEvent} from '@angular/material/chips';
+import {COMMA, ENTER, P} from '@angular/cdk/keycodes';
+
+import { ComponentService } from '../../services/component.service';
+import { DSMService } from '../../services/dsm.service';
+
+@Component({
+  selector: 'app-cohort-tag',
+  templateUrl: './cohort-tag.component.html',
+  styleUrls: ['./cohort-tag.component.scss']
+})
+export class CohortTagComponent implements OnInit {
+
+  @Input() ddpParticipantId: string;
+  @Input() dsm: Object;
+
+  addOnBlur = true;
+  readonly separatorKeysCodes = [ENTER, COMMA] as const;
+  tags: CohortTag[];
+
+  constructor(private compService: ComponentService, private dsmService: DSMService) { }
+
+  ngOnInit(): void {
+    if (this.hasNotCohortTag()) {
+      this.dsm['cohortTag'] = [];
+    }
+    this.tags = this.getTags();
+  }
+
+  private hasNotCohortTag(): boolean {
+    return !this.dsm['cohortTag'];
+  }
+
+  private getTags(): CohortTag[] {
+    return this.dsm['cohortTag'];
+  }
+
+  add(event: MatChipInputEvent): void {
+    const value = (event.value || '').trim();
+
+    if (value) {
+      const newTag = new CohortTag(value, this.ddpParticipantId);
+      this.dsmService.createCohortTag(JSON.stringify(newTag), this.compService.getRealm()).subscribe(cohortTagId => {
+        newTag.cohortTagId = parseInt(cohortTagId);
+      }, err => {
+        this.remove(newTag);
+      });
+      this.getTags().push(newTag);
+      // Clear the input value
+      event.chipInput!.clear();
```

Comment:
Just wondering what the `!` sign means here?

Reply:
@alextanlan to be honest I don't know, I just copied it from https://material.angular.io/components/chips/overview

Reply:
Interesting

Reply:
Interesting